CMakeLists.txt

```
@@ -814,6 +814,8 @@ if(NOT MSVC)
     string(APPEND CMAKE_CXX_FLAGS " -Wno-range-loop-analysis")
     string(APPEND CMAKE_CXX_FLAGS " -Wno-pass-failed")
   endif()
+  append_cxx_flag_if_supported("-Wmissing-prototypes" CMAKE_CXX_FLAGS)
+  append_cxx_flag_if_supported("-Wno-error=missing-prototypes" CMAKE_CXX_FLAGS)
```

Not sure what you are trying to achieve here. If We add `no-error` then one might not even be aware of the constraint and  free to ignore further regression.
```suggestion
  append_cxx_flag_if_supported("-Wmissing-prototypes" CMAKE_CXX_FLAGS)
```

---

I am afraid that the current code base is not error free for Wmissing-prototypes after this PR even when we have applied some modifications, so I decide to temporarily treat it as a warning only option. After several further PRs to gradually check and fix the violating files then we can finally enable it

---

I propose to restrict it to some files  or only to `clang-tidy`, as there is nothing more annoying then 100 of lines of warnings, and it encourages anti-pattern of introducing new ones. Right now build logs on Linux and Mac are really compact which makes finding new errors easy. Let's keep them that way.